kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -419,192 +401,180 @@ private LogSegment constructLogSegmentFromFileList(
-                                () -> {
-                                  final FileStatus lastDelta = deltas.get(deltas.size() - 1);
-                                  return FileNames.deltaVersion(new Path(lastDelta.getPath()));
-                                });
-
-                        return getLogSegmentWithMaxExclusiveCheckpointVersion(
-                                snapshotVersion, startCheckpoint)
-                            .orElseThrow(
-                                () ->
-                                    // No alternative found, but the directory contains files
-                                    // so we cannot return None.
-                                    new RuntimeException(
-                                        format(
-                                            "Checkpoint file to load version: %s is missing.",
-                                            startCheckpoint)));
-                      });
+    final Optional<CheckpointInstance> latestCompleteCheckpointOpt =
+        Checkpointer.getLatestCompleteCheckpointFromList(
+            listedCheckpointInstances, notLaterThanCheckpoint);
 
-                  return -1L;
-                });
-    logger.debug("newCheckpointVersion: {}", newCheckpointVersion);
+    if (!latestCompleteCheckpointOpt.isPresent() && startCheckpointVersionOpt.isPresent()) {
+      // In Step 1 we found a $startCheckpointVersion but now our LIST of the file system doesn't
+      // see it. This means that the checkpoint we thought should exist no longer does.
+      throw DeltaErrors.missingCheckpoint(tablePath.toString(), startCheckpointVersionOpt.get());
+    }
+
+    final long latestCompleteCheckpointVersion =
+        latestCompleteCheckpointOpt.map(x -> x.version).orElse(-1L);
+
+    logger.info("Latest complete checkpoint version: {}", latestCompleteCheckpointVersion);
+
+    /////////////////////////////////////////////////////////////////////////////////////////////
+    // Step 7: Grab all deltas in range [$latestCompleteCheckpointVersion + 1, $versionToLoad] //
+    /////////////////////////////////////////////////////////////////////////////////////////////
 
-    // TODO: we can calculate deltasAfterCheckpoint and deltaVersions more efficiently
-    // If there is a new checkpoint, start new lineage there. If `newCheckpointVersion` is -1,
-    // it will list all existing delta files.
     final List<FileStatus> deltasAfterCheckpoint =
-        deltas.stream()
+        listedDeltaFileStatuses.stream()
             .filter(
-                fileStatus ->
-                    FileNames.deltaVersion(new Path(fileStatus.getPath())) > newCheckpointVersion)
+                fs -> {
+                  final long deltaVersion = FileNames.deltaVersion(new Path(fs.getPath()));
+                  return latestCompleteCheckpointVersion + 1 <= deltaVersion
+                      && deltaVersion <= versionToLoadOpt.orElse(Long.MAX_VALUE);
```

Comment:
this is redundant right? since we only list up to versionToLoad? (for my understanding, it's fine to stay)

Reply:
yup! "redundant" assuming that the LIST call returns what we expect.